Add angular-google-maps as a Yarn dependency

luisramos0 opened the pull request:
#### What? Why?

Closes #5120

We have added npm and yarn to our toolset in openfoodfoundation/ofn-install#613.
All servers are now running yarn and on every deploy yarn install is executed to install dependencies.
No dependencies are currently installed with yarn (there are some **dev** dependencies in yarn but they are not deployed to staging or production servers).

This PR adds some npm/yarn dependencies and removes the dependencies JS code from our repository :tada:

From now on we dont need to add js dependencies to our repo, we can just use "yarn add" and require them in rails app/assets using their path on node_modules.

This was not working in rails 3 but is looking great in rails 4 :+1: 

Locally, developers will be able to keep using only npm except if they want to add a new JS dependency, they will need to install yarn locally.
I used this guide to install yarn locally and it just worked:
https://linuxize.com/post/how-to-install-yarn-on-ubuntu-18-04/
We need to add this to the dev guide somewhere after this PR is merged.

This didn't work for some CSS dependencies I tried to bring from spree_core Normalize and skeleton. These dependencies (fetched from their github repos), break the npm install deployment... a problem for another PR.

#### What should we test?
We should test angular-google-maps just like in #5121: check /maps is working and searching for enterprises by location in the /shops page is working.

#### Release notes
Added yarn to our toolset to manage JS dependencies (instead of copying the code of all JS dependencies to our repo).
Changelog Category: Changed

luisramos0 commented:
You dont need yarn to run the application afrer this. You only need yarn to add a new dependency or update an existing one. It's diferent.
Also, this doesnt require yarn in production. This doesnt need any change in ofn-install. Yarn here is only for development and git.

Adding Yarn in production and ofn-install would be a further step so that we take full advantage of yarn.

I think yarn adds value because yarn.lock will let us manage the exact dependencies in git. Think gems without Gemfile.lock.

From what I have read yarn is a standard for rails apps include latest rails where it works with webpacker. I am pretty sure this is the way to go.

What is the alternative @mkllnk? to use only npm because yarn is too much complexity?

kristinalim commented:
I like the direction of this PR. It really is a pain to manage JS library versions.

Agreeing with @mkllnk that yarn needs to be installed in the server and added to the deployment sequence. We compile assets in the server, so the node modules need to be installed in the server because they aren't stored in the Git repository.

BTW @luisramos0, I noticed that the bluebird version resolved by Yarn is different from what we are currently using. Are we using it anywhere else? It looks like it's a Promise library, and unless we know for sure that it's not used anywhere, we should probably lock it to the same version and upgrade in another PR for more comprehensive testing. If going with the upgrade, can you mention this in the release note?

luisramos0 commented:
ah, thanks @mkllnk for clarifying :pray: I used the wrong assumption that we were already using npm in production! But we are not.
That's why I said we wouldnt need to have yarn in prod, because I thought we already had npm.
Sorry, my mistake!

Now that I see we dont have npm, I think we should go straight to yarn, I think you agree about this.


Thanks Kristina for the heads up, it's a mistake in the PR. We only need bluebird because of angular-google-maps, you can see I have not added the bluebird dependency to package.json. I was planning to use the version that comes inside the angular-google-maps version under angular-google-maps/bower_components/bluebird/js/browser/bluebird.js which is 2.3.11 as we were using but I forgot to change the path in all.js.coffee.
I dont understand why yarn is resolving it to another version when angular-google-maps comes with 2.3.11!?
Anyway, I think it's easier in this case (and probably more correct because bluebird is not our direct dependency) to ignore yarn.lock and just import angular-google-maps/bower_components/bluebird/js/browser/bluebird.js directly.

Anyway, I need to work a little more on this PR and also on an ofn-install PR to get yarn to the servers :+1:   

Thanks

Matt-Yorkley commented:
I think I'd be a lot happier if we were using this to pull dependencies into the codebase in development and not relying on it in production...

luisramos0 commented:
@Matt-Yorkley that's like saying you dont trust bundler and you want to have the gems in the codebase, isnt it? I think that's because this is a new thing for us.
I think we should trust yarn to do it's job. I think we should do it in the environment just like we do bundle install, we will do yarn install.
What do other people think?

kristinalim commented:
I prefer not to include them in the repository.

Matt-Yorkley commented:
> that's like saying you dont trust bundler and you want to have the gems in the codebase, isnt it? I think that's because this is a new thing for us.

I'm just concerned about node-based package management specifically, as I've read multiple articles in the past about very serious problems with it. Like the maintainer closing or changing a repository and thousands of production environments suddenly failing, eg: https://www.zdnet.com/article/disgruntled-developer-breaks-thousands-of-javascript-node-js-apps/

I totally trust `bundler` in our production environment.

luisramos0 commented:
I think this step is the most natural thing to do.
Thousands of applications use npm and yarn as their package manager just like we use bundler.
From what I read this is the normal way of doing it in modern rails, we can add webpacker later:
https://gist.github.com/maxivak/2612fa987b9f9ed7cb53a88fcba247b3#npm-and-yarn

luisramos0 commented:
Moving again to Code Review as openfoodfoundation/ofn-install#613 is also Ready for Review.

mkllnk commented:
> The incident would happen if someone would also unpublish a popular package from rubygems, right?

Right. I somehow thought (hoped) that Rubygems would have a better policy around this but it looks like the author can just remove gems. And if they delete all gems of a namespace, that namespace is free to be used by anyone else straight away! We need to trust our gem authors...
https://help.rubygems.org/kb/gemcutter/removing-a-published-rubygem

> I think this step is the most natural thing to do.

I'm not sure. A lot of people go into the opposite direction: pre-compile all assets before releasing or building a whole container image. The idea is to make deployments as fail-safe and as re-producible as possible. If Rubygems is down, we can't deploy. That's not great but also not easy to get around. Getting around Npmjs is much easier.

I find this decision difficult. I'm weighing up between yarn generating vendor assets in development and yarn running on production. Let's make a pros and cons list:

Pros:

- Less foreign code in our repository.
- ???

Cons:

- An additional step for developers when first setting up the project and when dependencies change (like `bundle install` but without clear error message).
- An additional step during deployment.
- Deployments depend on Npmjs registry.

@luisramos0 Can you fill me in on the advantages? Something has to top the additional work for devs, especially newcomers which concerns me the most.

luisramos0 commented:
We have to consider the alternative: add node_modules to git :see_no_evil: and make the dev who is updating the dependency commit to node_modules.
It just sounds very conservative to me. I have difficulty considering why would we do it.

I think one important dimension here, and a positive thing about this change, is to make javascript and yarn first class citizens in this rails project.

The pros are more like cons of the alternative:
- it's a lot more code in the repo yes, that brings complexity and perceived  complexity to the project, removing all this JS code from the repo is a massive step in terms of simplifying the growing OFN codebase
- (maybe because I have worked in projects where javascript was the main language) having /node_modules in git sounds exactly like having /gems in git :see_no_evil: (/gems as in .rvm/gems/ruby-2.3.7/gems)

Regarding the cons you list:

>> An additional step for developers when first setting up the project and when dependencies change (like bundle install but without clear error message).

We already have that step for newcomers to be able to run specs locally:
https://github.com/openfoodfoundation/openfoodnetwork/blob/c38c59b0b9fdfa16209e2ebbc977f82f42219828/script/setup#L40-L42
Devs will have to run yarn install when these dependencies change (as much as they now have to run it when the current karma/devDependencies change). It will be a lot more infrequent than Gemfile changes.

Maybe we can have a specific script in the repo to run yarn install and bundle install together.

>> An additional step during deployment.

It's a script doing it, no extra work required. If you mean the risk of failure an extra step adds: npmjs registry is trusted by so many orgs, it will not fail. And yarn is used by so many people, it is deterministic. I think this is very low risk.

>> Deployments depend on Npmjs registry.

imo this is such a low risk that it doesnt even qualify as a con. It's like saying we would be better off if we didnt dependend on rubygems and we considered copying all ruby gems to github.

### Decision
Looks like @mkllnk and @Matt-Yorkley are tending to have node_modules in the repo.
And me and @kristinalim to trust npm registry and yarn to do their jobs
@sauloperez thoughts? :-)

luisramos0 commented:
I see that rails has a .gitignore node_modules here:
https://github.com/rails/rails/blob/b7c846afafd82865948e4db939348e61bba47368/.gitignore#L14

sauloperez commented:
> I think this step is the most natural thing to do.

I think so too.

I feel like the lack of experience and fear is making us overengineer things a bit. As @luisramos0 and @kristinalim stated, this is not different than Ruby and Bundler: bundler => yarn, rubygems => NPM registry, Gemfile.lock => yarn.lock, vendor/bundle/ => node_modules/.

What you mention about the JS-gate where some libraries got removed, that has nothing to do with yarn. Blame it on JS, but we can't change it #shitHappens.

Doesn't storing the dependencies in the repo beat the purpose of a package manager? I thought the idea was to overcome the 2000s practice of having a `vendor/jquery.min.js` and build all assets ourselves.

Personally, I'm pretty confident with yarn as my experience with it so far it's been flawless. I really see all this is for the better.

sauloperez commented on config/initializers/assets.rb:
```diff
@@ -0,0 +1 @@
+Rails.application.config.assets.paths << Rails.root.join('node_modules')
```
this means we'll also include any dev dependencies. Given this is the first step, it's fine as long as we are all aware of it.

luisramos0 replied:
only in dev.
ofn-install task will run in production mode:
https://github.com/openfoodfoundation/ofn-install/pull/613/files#diff-a42145ed13cdec37c3b3553e3b3b9169R82

you can also do it locally with
yarn install --production

luisramos0 replied:
you can test this with
yarn install --production --purge

mkllnk commented:
We have a majority. 3 devs for yarn packing during deploy and 2 devs for yarn packaging during development. That's fine, no veto from me. Let's do it.

luisramos0 commented:
Thanks Maikel!

I think there's no need to add risk to v3 (rails 4) deploy with this change. So I'll hold this PR until rails 4 is live, and then we can go ahead with this.
Meanwhile we can get ofn-install/613 live.

luisramos0 commented:
I got into some trouble with angular-google-maps because it calls bower install in a post install script and bower is not necessarily available in the environment (bower is another js dep management tool). We can leave this one for later.

I switched to momentjs and moved it from Gemfile to yarn: I managed to upgrade it out of the box :+1: 
PR is ready for review here: https://github.com/openfoodfoundation/openfoodnetwork/pull/6049
